src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,198 @@
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute < value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpLessThanEquals: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute <= value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpGreaterThan: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute > value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+          case ScanType::OpGreaterThanEquals: {
+            auto value = typed_dict_segment->find_in_dict(search_value());
+            for (auto index = ValueID(0); index < size; index++) {
+              auto attribute = typed_dict_segment->attribute_vector()->get(i);
+              if (attribute >= value) {
+                posList.emplace_back(RowID({i, ChunkOffset(index)}));
+              }
+            }
+            break;
+          }
+        }
+
+      const auto typed_ref_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+      if(typed_ref_segment != nullptr) {
+        auto pos_list_size = typed_ref_segment->pos_list()->size();
+
+        for(size_t index = 0; index < pos_list_size; index++) {
+          auto row_id = typed_ref_segment->pos_list()->at(index);
+          auto value = typed_ref_segment
+            ->referenced_table()
+            ->get_chunk(row_id.chunk_id)
+            .get_segment(typed_ref_segment->referenced_column_id())
+            ->operator[](row_id.chunk_offset);
+
+          if(
+              comparator(
+                scan_type(),
+                type_cast<Type>(value),
+                type_cast<Type>(search_value())
+              )
+            ) {
+              posList.emplace_back(RowID{i, ChunkOffset(index)});
+            };
+        }
+      }
+      }
+    });
+  }
+
+  auto& chunk = out_table->get_chunk(ChunkID{0});
+
+  for (auto column = ColumnID(0); column < in_table->column_count(); column++) {
+    out_table->add_column_definition(in_table->column_name(column), in_table->column_type(column));
+  }
+
+  for (auto column = ColumnID(0); column < _input_table_left()->column_count(); column++) {
+    const auto shared_segment = std::make_shared<ReferenceSegment>(in_table, column, std::make_shared<PosList>(posList));
```

Comment:
(Also, you could use different shared pointers to the same `PosList` instance here, but your current code will make a copy for each ReferenceSegment)

Reply:
> This means that this will break if different columns have different levels of indirection (one columns is a dictionary or value segment, one is a reference segment)

For the sake of completeness: This is currently not allowed in Hyrise

> [imagine] the result of a join with two columns A and B,

Yepp, valid point, but out-of-scope for this sprint as we haven't really talked about joins yet. If you implemented that, I'll give you some extra points.

> (Also, you could use different shared pointers to the same PosList instance here, but your current code will make a copy for each ReferenceSegment)

Good point